client/ipa-join.c

```diff
@@ -612,8 +617,246 @@ join_krb5(const char *ipaserver, char *hostname, char **hostdn, const char **pri
+    headers = curl_slist_append(headers, user_agent);
+
+    headers = curl_slist_append(headers, "Accept: application/json");
+    headers = curl_slist_append(headers, "Content-Type: application/json");
+    CURL_SETOPT(curl, CURLOPT_HTTPHEADER, headers);
+
+    CURL_SETOPT(curl, CURLOPT_CAINFO, DEFAULT_CA_CERT_FILE);
+
+    CURL_SETOPT(curl, CURLOPT_WRITEFUNCTION, &jsonrpc_handle_response);
+    CURL_SETOPT(curl, CURLOPT_WRITEDATA, &cb);
+
+    /* delegate authentication to GSSAPI */
+    CURL_SETOPT(curl, CURLOPT_GSSAPI_DELEGATION, CURLGSSAPI_DELEGATION_FLAG);
+    CURL_SETOPT(curl, CURLOPT_HTTPAUTH, CURLAUTH_NEGOTIATE);
+    CURL_SETOPT(curl, CURLOPT_USERPWD, ":");
+
+    if (debug)
+        CURL_SETOPT(curl, CURLOPT_VERBOSE, 1L);
+
+    /* create the JSON-RPC payload */
+    json_post = json_pack_ex(&j_error, 0, "{s:s, s:[[s], {s:s, s:s}]}", "method", "join", "params", host, "nsosversion",
+                             uinfo.release, "nshardwareplatform", uinfo.machine);
+
+    if (json_post == NULL) {
+        if (debug)
+            fprintf(stderr, _("json_pack failed: %s\n"), j_error.text);
+
+        rval = 17;
+        goto cleanup;
+    }
+
+    json_post_str = json_dumps(json_post, 0);
+    CURL_SETOPT(curl, CURLOPT_POSTFIELDS, json_post_str);
+
+    if (debug)
+        fprintf(stderr, _("JSON-RPC request:\n%s\n"), json_post_str);
+
+    /* Perform the call and check for errors */
+    res = curl_easy_perform(curl);
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &resp_code);
+    if (res != CURLE_OK)
+    {
+        if (debug)
+            fprintf(stderr, _("JSON-RPC call failed with curl error: %s\n"), curl_easy_strerror(res));
+
+        rval = 17;
+        goto cleanup;
+    }
+
+    if (resp_code != 200) {
+        if (debug)
+            fprintf(stderr, _("JSON-RPC call failed with status code: %li\n"), resp_code);
+
+        if (!quiet && resp_code == 401)
+            fprintf(stderr, _("JSON-RPC call was unauthorized. Check your credentials.\n"));
+
+        rval = 17;
+        goto cleanup;
+    }
+
+    if (cb.payload != NULL) {
+        if (debug) {
+            fprintf(stderr, _("JSON-RPC response:\n%s\n"), cb.payload);
+        }
+    }
+
+    j_root = json_loads(cb.payload, 0, &j_error);
+    if (j_root == NULL) {
+        if (debug)
+            fprintf(stderr, _("Parsing JSON-RPC response failed: %s\n"), j_error.text);
+
+        rval = 17;
+        goto cleanup;
+    }
+
+    j_result = json_object_get(j_root, "result");
+
+    char* tmp_hostdn;
+    if (json_unpack_ex(j_result, &j_error, JSON_STRICT, "[s, o]", &tmp_hostdn, &j_props) != 0) {
+        if (debug)
+            fprintf(stderr, _("Parsing JSON-RPC response failed: %s\n"), j_error.text);
+
+        rval = 17;
+        goto cleanup;
+    }
+    ASPRINTF((char **)hostdn, "%s", tmp_hostdn);
+
+    char* tmp_princ;
+    if (json_unpack_ex(json_object_get(j_props, "krbprincipalname"), &j_error, JSON_STRICT, "[s]", &tmp_princ) != 0) {
```

Comment:
This is unsafe, ``(json_object_get(j_props, "krbprincipalname")`` can fail and return NULL.